presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1411,48 +1414,82 @@ else if (expression instanceof LongLiteral) {
         return orderByExpressions;
     }
 
-    private List<FieldOrExpression> analyzeGroupBy(QuerySpecification node, RelationType tupleDescriptor, AnalysisContext context, List<FieldOrExpression> outputExpressions)
+    private List<List<FieldOrExpression>> analyzeGroupBy(QuerySpecification node, RelationType tupleDescriptor, AnalysisContext context, List<FieldOrExpression> outputExpressions)
     {
-        ImmutableList.Builder<FieldOrExpression> groupByExpressionsBuilder = ImmutableList.builder();
-        if (!node.getGroupBy().isEmpty()) {
-            // Translate group by expressions that reference ordinals
-            for (Expression expression : node.getGroupBy()) {
-                // first, see if this is an ordinal
-                FieldOrExpression groupByExpression;
+        if (node.getGroupBy().isEmpty()) {
+            analysis.setGroupingSets(node, ImmutableList.of());
+            return ImmutableList.of();
+        }
 
-                if (expression instanceof LongLiteral) {
-                    long ordinal = ((LongLiteral) expression).getValue();
-                    if (ordinal < 1 || ordinal > outputExpressions.size()) {
-                        throw new SemanticException(INVALID_ORDINAL, expression, "GROUP BY position %s is not in select list", ordinal);
-                    }
+        ImmutableList.Builder<List<FieldOrExpression>> groupingSetsBuilder = ImmutableList.builder();
+        List<GroupingElement> groupingElements = node.getGroupBy().stream()
```

Comment:
I don't think it matters. It's equivalent to:

`S1 x S2 x S3`,  where

```
S1 = { (a,b) }
S2 = { (c,d) }
S3 = { (e,f), (e), (f), () }
```

As we discussed the other day, the following forms should be equivalent:

```
GROUP BY 
    GROUPING SETS ((a)),
    GROUPING SETS ((b)),
    GROUPING SETS ((c))
```
```
GROUP BY a, b, c
```
```
GROUP BY (a, b), c
```